src/main/java/jcifs/spnego/NegTokenTarg.java

```diff
@@ -0,0 +1,152 @@
+
+import org.bouncycastle.asn1.ASN1EncodableVector;
+import org.bouncycastle.asn1.ASN1Enumerated;
+import org.bouncycastle.asn1.ASN1InputStream;
+import org.bouncycastle.asn1.ASN1ObjectIdentifier;
+import org.bouncycastle.asn1.ASN1OctetString;
+import org.bouncycastle.asn1.ASN1Sequence;
+import org.bouncycastle.asn1.ASN1TaggedObject;
+import org.bouncycastle.asn1.DEROctetString;
+import org.bouncycastle.asn1.DEROutputStream;
+import org.bouncycastle.asn1.DERSequence;
+import org.bouncycastle.asn1.DERTaggedObject;
+
+
+@SuppressWarnings ( "javadoc" )
+public class NegTokenTarg extends SpnegoToken {
+
+    public static final int UNSPECIFIED_RESULT = -1;
+    public static final int ACCEPT_COMPLETED = 0;
+    public static final int ACCEPT_INCOMPLETE = 1;
+    public static final int REJECTED = 2;
+    public static final int REQUEST_MIC = 3;
+
+    private ASN1ObjectIdentifier mechanism;
+
+    private int result = UNSPECIFIED_RESULT;
+
+
+    public NegTokenTarg () {}
+
+
+    public NegTokenTarg ( int result, ASN1ObjectIdentifier mechanism, byte[] mechanismToken, byte[] mechanismListMIC ) {
+        setResult(result);
+        setMechanism(mechanism);
+        setMechanismToken(mechanismToken);
+        setMechanismListMIC(mechanismListMIC);
+    }
+
+
+    public NegTokenTarg ( byte[] token ) throws IOException {
+        parse(token);
+    }
+
+
+    public int getResult () {
+        return this.result;
+    }
+
+
+    public void setResult ( int result ) {
+        this.result = result;
+    }
+
+
+    public ASN1ObjectIdentifier getMechanism () {
+        return this.mechanism;
+    }
+
+
+    public void setMechanism ( ASN1ObjectIdentifier mechanism ) {
+        this.mechanism = mechanism;
+    }
+
+
+    @Override
+    public byte[] toByteArray () {
+        try {
+            ByteArrayOutputStream collector = new ByteArrayOutputStream();
+            DEROutputStream der = new DEROutputStream(collector);
+            ASN1EncodableVector fields = new ASN1EncodableVector();
+            int res = getResult();
+            if ( res != UNSPECIFIED_RESULT ) {
+                fields.add(new DERTaggedObject(true, 0, new ASN1Enumerated(res)));
+            }
+            ASN1ObjectIdentifier mech = getMechanism();
+            if ( mech != null ) {
+                fields.add(new DERTaggedObject(true, 1, mech));
+            }
+            byte[] mechanismToken = getMechanismToken();
+            if ( mechanismToken != null ) {
+                fields.add(new DERTaggedObject(true, 2, new DEROctetString(mechanismToken)));
+            }
+            byte[] mechanismListMIC = getMechanismListMIC();
+            if ( mechanismListMIC != null ) {
+                fields.add(new DERTaggedObject(true, 3, new DEROctetString(mechanismListMIC)));
+            }
+            der.writeObject(new DERTaggedObject(true, 1, new DERSequence(fields)));
+            return collector.toByteArray();
+        }
+        catch ( IOException ex ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)